File: 85-maximal-rectangle/maximal-rectangle.java
import java.util.*;

class Solution {
    public int maximalRectangle(char[][] matrix) {

        int rows = matrix.length;
        int cols = matrix[0].length;

        int[] heights = new int[cols + 1];

        int ans = 0;

        for (int r = 0; r < rows; r++) {

            // Build histogram
            for (int c = 0; c < cols; c++) {

                if (matrix[r][c] == '1') {
                    heights[c]++;
                } else {
                    heights[c] = 0;
                }
            }

            Stack<Integer> stack = new Stack<>();

            for (int i = 0; i <= cols; i++) {

                while (!stack.isEmpty() &&
                       heights[stack.peek()] > heights[i]) {

                    int h = heights[stack.pop()];

                    int width;

                    if (stack.isEmpty()) {
                        width = i;
                    } else {
                        width = i - stack.peek() - 1;
                    }

                    ans = Math.max(ans, h * width);
                }

                stack.push(i);
            }
        }

        return ans;
    }
}

File: 85-maximal-rectangle/maximal-rectangle.py
class Solution:
    def maximalRectangle(self, matrix):

        if not matrix:
            return 0

        cols = len(matrix[0])
        heights = [0] * (cols + 1)

        ans = 0

        for row in matrix:

            # Build histogram
            for i in range(cols):
                if row[i] == "1":
                    heights[i] += 1
                else:
                    heights[i] = 0

            stack = []

            for i in range(cols + 1):

                while stack and heights[stack[-1]] > heights[i]:

                    h = heights[stack.pop()]

                    if stack:
                        width = i - stack[-1] - 1
                    else:
                        width = i

                    ans = max(ans, h * width)

                stack.append(i)

        return ans